aten/src/ATen/native/utils/ParamsHash.h

```diff
@@ -38,5 +38,57 @@ struct ParamsEqual {
   }
 };
 
+// Provide explicit byte-for-byte constructors to avoid uwittingly leaving
+// padding bytes unitialized (e.g., when passing Params by value)
+template <typename T>
+struct ParamsWrapper {
+  T pod;
+  static_assert(std::is_standard_layout_v<T>, "ParamsWrapper cannot wrap non-POD data");
+
+  ParamsWrapper() {
+    memset(&(this->pod), 0, sizeof(T));
+  }
+
+  ParamsWrapper(const ParamsWrapper<T> &other) {
+    memcpy(&(this->pod), &(other.pod), sizeof(T));
+  }
+
+  ParamsWrapper(ParamsWrapper<T> &&other) {
+    memcpy(&(this->pod), &(other.pod), sizeof(T));
+  }
+};
+
+// Wrapped version: this allows the outer struct to have custom copy and move
+// constructors for additional safety
+template <typename ParamsWrapper>
+struct ParamsWrapperHash {
+  // Params must be a POD because we read out its memory
+  // contents as char* when hashing
+  static_assert(std::is_standard_layout_v<decltype(ParamsWrapper::pod)>, "ParamsWrapper cannot wrap non-POD data");
+
+  size_t operator()(const ParamsWrapper& params_wrapper) const {
+    auto ptr = reinterpret_cast<const uint8_t*>(&(params_wrapper.pod));
+    uint32_t value = 0x811C9DC5;
+    for (const auto i : c10::irange(sizeof(params_wrapper.pod))) {
+      value ^= ptr[i];
+      value *= 0x01000193;
+    }
+    return (size_t)value;
+  }
+};
+
+template <typename ParamsWrapper>
+struct ParamsWrapperEqual {
```

Comment:
I don't understand why `ParamsEqual` existed in the first place. Doesn't the default `operator==` work? The default one should be a item-by-item comparison, instead of a byte-by-byte one?

Reply:
I agree, and will check if the default operator would work.